OpenSim/Simulation/Model/Model.cpp

```diff
@@ -1771,26 +1774,28 @@ Vector& Model::updControls(const SimTK::State &s) const
 
 void Model::markControlsAsValid(const SimTK::State& s) const
 {
+  std::unique_lock<std::mutex> lock(*controlsCacheLock);
     if( (!_system) || (!_modelControlsIndex.isValid()) ){
-        throw Exception("Model::markControlsAsValid() requires an initialized Model./n" 
+        throw Exception("Model::markControlsAsValid() requires an initialized Model./n"
             "Prior call to Model::initSystem() is required.");
     }
 
-    Measure_<Vector>::Result controlsCache = 
+    Measure_<Vector>::Result controlsCache =
         Measure_<Vector>::Result::getAs(_system->updDefaultSubsystem()
             .getMeasure(_modelControlsIndex));
     controlsCache.markAsValid(s);
 }
 
 void Model::setControls(const SimTK::State& s, const SimTK::Vector& controls) const
-{   
+{
+  std::unique_lock<std::mutex> lock(*controlsCacheLock);
```

Comment:
The lock prevents other threads from calling any of the other control-related methods in Model if another thread is calling a control-related method. Since the controls cache is shared, if multiple threads try to write to the controls cache it'll cause a data race. Although the threads aren't calling these control-related methods when realizing dynamics (when the program launches separate threads), I wanted to add locks just as a precaution. Do you think this is necessary or should they just be added on a as-needed basis?